src/NuGet.Core/NuGet.Configuration/Settings/SettingsLoadingContext.cs

```diff
@@ -47,15 +46,19 @@ internal SettingsFile GetOrCreateSettingsFile(string filePath, bool isMachineWid
                 }
 
                 var file = new FileInfo(filePath);
-                settingsFile = new Lazy<SettingsFile>(() => new SettingsFile(file.DirectoryName, file.Name, isMachineWide, isReadOnly));
+
```

Comment:
@nkolev92 I had to update `SettingsLoadingContext` here because it was caching Exceptions since the `Lazy<T>` was always added to the list.  The new logic only adds it if accessing the `Value` doesn't throw an exception which still happens in a `Semaphore`.  Do you have any concerns about it?

Reply:
What's the net outcome? 


What's different about the following scenarios: 

* 2 projects being restored, with the same unparsable config in the path.
* 2 projects with different customer SDKs being restored, with the same unparsable config in the path.

Reply:
> 2 projects being restored, with the same unparsable config in the path.

At the moment they both throw, one is the real exception and the other is cached but the exception doesn't come from accessing the `Lazy<T>`, it comes from accessing the `List<T>` here:

https://github.com/NuGet/NuGet.Client/blob/493d6b38484377e8c4d713d960097b97ce3ba4e9/src/NuGet.Core/NuGet.Configuration/Settings/SettingsLoadingContext.cs#L42

I found that for a `SettingsLoadingContext`, once any settings throw an exception, then any access to the settings context throws because it walk through the ones that were added even if they threw.  Generally this is okay because the restore failed and the settings context is not re-used.  But in the context of our unit test, the settings context is re-used after trying to access a bad config.

> 2 projects with different customer SDKs being restored, with the same unparsable config in the path.
This would throw multiple times like above, but since the restore failed the user wouldn't really notice that its throwing again.  As far as I can tell, only our unit tests are affected

Reply:
Thanks for the explanation. Makes sense to me.

Reply:
I've decided to improve SettingsLoadingContext in a different commit since its something I've been wanting to do for a while and I don't want to pollute this change in case either need a revert...